src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,94 @@
+package com.app.onlinebookstore.service.impl;
+
+import com.app.onlinebookstore.dto.shoppingcart.CartItemDto;
+import com.app.onlinebookstore.dto.shoppingcart.CreateCartItemDto;
+import com.app.onlinebookstore.dto.shoppingcart.ShoppingCartDto;
+import com.app.onlinebookstore.dto.shoppingcart.UpdateCartItemDto;
+import com.app.onlinebookstore.exception.EntityNotFoundException;
+import com.app.onlinebookstore.mapper.CartItemMapper;
+import com.app.onlinebookstore.mapper.ShoppingCartMapper;
+import com.app.onlinebookstore.model.Book;
+import com.app.onlinebookstore.model.CartItem;
+import com.app.onlinebookstore.model.ShoppingCart;
+import com.app.onlinebookstore.repository.BookRepository;
+import com.app.onlinebookstore.repository.CartItemRepository;
+import com.app.onlinebookstore.repository.ShoppingCartRepository;
+import com.app.onlinebookstore.service.ShoppingCartService;
+import com.app.onlinebookstore.service.UserService;
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final UserService userService;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto find() {
+        return shoppingCartMapper.toDto(shoppingCartRepository
+                .findByUserId(userService.getUser().getId())
+                .orElseGet(this::createShoppingCart));
+    }
+
+    @Override
+    @Transactional
+    public CartItemDto save(CreateCartItemDto createCartItemDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findByUserId(userService.getUser().getId())
+                .orElseGet(this::createShoppingCart);
+        CartItem cartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(createCartItemDto.getBookId()))
+                .peek(item -> {
+                    item.setQuantity(createCartItemDto.getQuantity());
+                    item.setDeleted(false);
+                })
+                .findFirst().orElseGet(() -> createCartItem(createCartItemDto));
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto update(UpdateCartItemDto updateCartItemDto, Long id) {
```

Comment:
what will be if user will specify id of item which doesn't belong to him? I think this case also should be validated

Reply:
If it happens the program return error because in repository i searching by userid and this id, so i don't think that it needed to be validated